File: backend/app/routers/restaurants.py
"""restaurant_brand endpoints. See docs/API_CONTRACTS.md "Restaurants
(restaurant_brand)".

Public: GET /restaurants/{id}, GET /restaurants/{id}/locations
(backend/CLAUDE.md "Public Routes"). Everything else requires auth,
including the owner/admin-scoped list below (bare GET /restaurants).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.auth import (
    CurrentUser,
    get_current_user_optional,
    require_admin,
    require_brand_write_access,
    require_owner,
    require_owner_or_admin,
    require_registered_user,
)
from app.dependencies.db import get_db
from app.dependencies.pagination import Pagination, pagination_params
from app.schemas.follow import FollowOut
from app.schemas.restaurant import (
    LocationListResponse,
    RestaurantCreate,
    RestaurantListResponse,
    RestaurantListStatusValue,
    RestaurantOut,
    RestaurantSortValue,
    RestaurantUpdate,
)
from app.services import follow_service, location_service, restaurant_service

router = APIRouter(prefix="/restaurants", tags=["restaurants"])


@router.get("", response_model=RestaurantListResponse)
async def list_restaurants(
    owner_id: int | None = Query(
        default=None,
        description="Admin only — ignored for an owner caller, who is always "
        "filtered to their own owner_id regardless of this param.",
    ),
    owner_email: str | None = Query(
        default=None,
        max_length=255,
        description="Admin only — same ignored-for-owner-caller rule as "
        "owner_id. Case-insensitive substring match against the brand "
        "owner's owner_account.email. A brand with no owner (unclaimed, "
        "owner_id IS NULL) never matches a non-empty owner_email filter.",
    ),
    name: str | None = Query(
        default=None,
        max_length=255,
        description="Admin only. Case-insensitive substring match against "
        "restaurant_brand.name.",
    ),
    location_status: RestaurantListStatusValue | None = Query(
        default=None,
        alias="status",
        description="Admin only. Matches a brand if ANY of its locations "
        "currently has this restaurant_location.status value. The "
        "pseudo-value `deleted` instead returns ONLY soft-deleted "
        "listings; every other request (and every owner request) excludes "
        "soft-deleted listings.",
    ),
    is_paid: bool | None = Query(
        default=None,
        description="Admin only. Matches a brand if ANY of its locations "
        "has this restaurant_location.is_paid value (true = at least one "
        "paid location, false = at least one free location).",
    ),
    city: str | None = Query(
        default=None,
        max_length=120,
        description="Admin only. Case-insensitive exact match against "
        "restaurant_location.city. Matches a brand if ANY of its "
        "locations is in that city — a brand can have locations across "
        "multiple cities.",
    ),
    is_claimed: bool | None = Query(
        default=None,
        description="Admin only. Matches restaurant_brand.is_claimed exactly.",
    ),
    sort: RestaurantSortValue | None = Query(
        default=None,
        description="Optional. Omitted keeps the default id-ascending order. "
        "`followers` sorts by follower_count descending (ties broken by id "
        "ascending) — available to the same owner/admin caller as the rest "
        "of this endpoint, not admin-only.",
    ),
    pagination: Pagination = Depends(pagination_params),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_owner_or_admin),
) -> RestaurantListResponse:
    """Auth: owner or admin. Owner caller: hard-filtered server-side to
    their own brands, no query param can widen this. Admin caller: the
    optional `owner_id` query param, omitted returns all brands
    (docs/API_CONTRACTS.md "GET /restaurants").

    `owner_email`/`name`/`status`/`is_paid`/`city`/`is_claimed` (added for
    the admin listings management page, docs/API_CONTRACTS.md "GET
    /restaurants" filters) follow the exact same admin-only,
    silently-ignored-for-an-owner-caller rule as `owner_id` — see
    `restaurant_service.list_restaurants`. All provided filters combine
    with AND (docs/API_CONTRACTS.md "GET /search" facet convention —
    independent filters AND together; there is no OR-within-a-filter case
    here since every new param here is single-valued, unlike `/search`'s
    `cuisine[]`/`dietary[]`/`type[]`).

    `sort` (added for the admin listings "Most followed" sort control) is
    NOT admin-only, unlike the filters above — see
    `restaurant_service.list_restaurants`'s docstring.
    """
    return await restaurant_service.list_restaurants(
        db,
        current_user,
        owner_id,
        pagination,
        owner_email=owner_email,
        name=name,
        status=location_status,
        is_paid=is_paid,
        city=city,
        is_claimed=is_claimed,
        sort=sort,
    )


@router.get("/{id_or_slug}", response_model=RestaurantOut)
async def get_restaurant(id_or_slug: str, db: AsyncSession = Depends(get_db)) -> RestaurantOut:
    """`id_or_slug` may be the numeric `restaurant_brand.id` or its `slug`
    (docs/API_CONTRACTS.md "GET /restaurants/{id}")."""
    return await restaurant_service.get_restaurant(db, id_or_slug)


@router.get("/{brand_id}/locations", response_model=LocationListResponse)
async def list_restaurant_locations(
    brand_id: int,
    pagination: Pagination = Depends(pagination_params),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser | None = Depends(get_current_user_optional),
) -> LocationListResponse:
    """Public by default — active locations only (backend/CLAUDE.md
    "Public Routes"; docs/API_CONTRACTS.md "GET /restaurants/{id}/locations").
    An authenticated caller who owns this brand, or an admin, additionally
    sees their own deactivated locations here (docs/PROJECT_PLAN.csv
    "Serialize paid_until/is_active on location endpoints + let owner see
    own deactivated locations") — every other caller (anonymous, a
    manager, a registered_user, or an owner who does not own this brand)
    is unaffected. See `location_service.list_locations_for_brand` /
    `_caller_may_see_inactive_locations` for the exact rule.
    """
    return await location_service.list_locations_for_brand(
        db, brand_id, pagination, current_user
    )


@router.post("", response_model=RestaurantOut, status_code=status.HTTP_201_CREATED)
async def create_restaurant(
    body: RestaurantCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_owner),
) -> RestaurantOut:
    return await restaurant_service.create_restaurant(db, body, current_user)


@router.patch("/{brand_id}", response_model=RestaurantOut)
async def update_restaurant(
    brand_id: int,
    body: RestaurantUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_brand_write_access),
) -> RestaurantOut:
    return await restaurant_service.update_restaurant(db, brand_id, body, current_user)


@router.delete("/{brand_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def delete_restaurant(
    brand_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> None:
    """Admin only. SOFT delete: stamps `restaurant_brand.deleted_at` and
    deactivates every active location of the brand in the same transaction.
    Idempotent, never 409s (docs/API_CONTRACTS.md "DELETE /restaurants/{id}")."""
    await restaurant_service.delete_restaurant(db, brand_id, current_user)


@router.post("/{brand_id}/restore", response_model=RestaurantOut)
async def restore_restaurant(
    brand_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> RestaurantOut:
    """Admin only. Clears `deleted_at`; the brand's locations stay
    deactivated until re-enabled via `POST /locations/{id}/status`."""
    return await restaurant_service.restore_restaurant(db, brand_id, current_user)


@router.post("/{brand_id}/follow", response_model=FollowOut)
async def follow_restaurant(
    brand_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_registered_user),
) -> FollowOut:
    return await follow_service.follow_brand(db, brand_id, current_user)


@router.delete("/{brand_id}/follow", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def unfollow_restaurant(
    brand_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_registered_user),
) -> None:
    await follow_service.unfollow_brand(db, brand_id, current_user)

File: frontend/src/lib/api/restaurants.ts
// Typed client for /restaurants (restaurant_brand) — docs/API_CONTRACTS.md
// "Restaurants (`restaurant_brand`)".
import { apiFetch, toQueryString } from "./client";
import type { PaginatedResponse, PaginationParams } from "@/types/common";
import type { FollowOut } from "@/types/follow";
import type { LocationSummary } from "@/types/location";
import type { ListingStatusFilter } from "@/lib/adminListings";
import type {
  CreateRestaurantInput,
  LocationPage,
  RestaurantBrand,
  RestaurantPublic,
  UpdateRestaurantInput,
} from "@/types/restaurant";

/** Admin-only filters on `GET /restaurants` (docs/API_CONTRACTS.md "GET
 * /restaurants" filters, added alongside the `/admin/listings` filter
 * bar) — every field here is silently ignored server-side for a
 * non-admin (owner) caller, same as `ownerId` above. All provided
 * filters combine with AND. */
export interface AdminRestaurantListFilters {
  ownerId?: number;
  /** Case-insensitive substring match against the owner's email. */
  ownerEmail?: string;
  /** Case-insensitive substring match against the brand name. */
  name?: string;
  /** Matches a brand if ANY of its locations has this status. The
   * pseudo-value `"deleted"` instead lists ONLY soft-deleted listings
   * (every other request excludes them). */
  status?: ListingStatusFilter;
  /** Matches a brand if ANY of its locations has this is_paid value. */
  isPaid?: boolean;
  /** Case-insensitive exact match; matches a brand if ANY of its
   * locations is in this city. */
  city?: string;
  isClaimed?: boolean;
}

/** `GET /restaurants` `sort` query param — NOT admin-only, unlike
 * `AdminRestaurantListFilters` above (docs/API_CONTRACTS.md "GET
 * /restaurants"). Omitted keeps the existing default order unchanged. */
export type RestaurantSort = "followers";

/**
 * GET /restaurants — auth: owner or admin (docs/API_CONTRACTS.md, added
 * 2026-09-13 to unblock the owner portal dashboard). Owner caller: always
 * implicitly filtered to their own `owner_id` server-side — no query param
 * can widen this. Admin caller: optional `ownerId` filter, omitted returns
 * all brands. `AdminRestaurantListFilters`' other fields are likewise
 * admin-only.
 *
 * FLAGGED CONTRACT GAP (see this PR's description): there is no manager
 * path for this endpoint at all — "Auth: owner or admin" only. A manager
 * caller gets a 403, so this function cannot be used to list a manager's
 * assigned locations; the dashboard page handles that role separately.
 */
export async function getMyRestaurants(
  params: PaginationParams & AdminRestaurantListFilters & { sort?: RestaurantSort } = {},
  accessToken: string
): Promise<PaginatedResponse<RestaurantBrand>> {
  const query = toQueryString({
    owner_id: params.ownerId,
    owner_email: params.ownerEmail,
    name: params.name,
    status: params.status,
    is_paid: params.isPaid,
    city: params.city,
    is_claimed: params.isClaimed,
    sort: params.sort,
    page: params.page,
    page_size: params.page_size,
  });

  return apiFetch<PaginatedResponse<RestaurantBrand>>(
    `/restaurants${query}`,
    { method: "GET" },
    { accessToken }
  );
}

/** GET /restaurants/{id} — public. Used by the SSR listing page. */
export async function getRestaurantById(id: number): Promise<RestaurantBrand> {
  return apiFetch<RestaurantBrand>(
    `/restaurants/${id}`,
    { method: "GET" },
    { revalidateSeconds: 60 }
  );
}

/**
 * GET /restaurants/{id_or_slug} — the brand only (no locations). The public
 * restaurant pages use `getRestaurantPublicBySlug` / `getLocationPageBySlugs`
 * below instead (brand + locations in one round trip); this remains for the
 * brand-level "report a problem" page. `{id}` resolves a slug too
 * (docs/API_CONTRACTS.md "GET /restaurants/{id}").
 */
export async function getRestaurantBySlug(slug: string): Promise<RestaurantBrand> {
  return apiFetch<RestaurantBrand>(
    `/restaurants/${slug}`,
    { method: "GET" },
    { revalidateSeconds: 60 }
  );
}

/**
 * GET /restaurants/by-slug/{brand_slug} — public. The brand plus its ACTIVE
 * locations (cards) in ONE round trip: what `/restaurant/{brandSlug}` renders
 * from (one location -> that location's profile, several -> the landing page).
 * 404 for an unknown or soft-deleted brand.
 */
export async function getRestaurantPublicBySlug(brandSlug: string): Promise<RestaurantPublic> {
  return apiFetch<RestaurantPublic>(
    `/restaurants/by-slug/${encodeURIComponent(brandSlug)}`,
    { method: "GET" },
    { revalidateSeconds: 60 }
  );
}

/**
 * GET /restaurants/by-slug/{brand_slug}/locations/{location_slug} — public and
 * viewer-aware, one round trip (brand + full location). Same visibility and
 * deal-content gating as `GET /locations/{id}`: pass `accessToken` when signed
 * in (deal content; an owner/admin/manager previewing a hidden location),
 * omit it for the anonymous, 60s-cached read.
 */
export async function getLocationPageBySlugs(
  brandSlug: string,
  locationSlug: string,
  accessToken?: string
): Promise<LocationPage> {
  return apiFetch<LocationPage>(
    `/restaurants/by-slug/${encodeURIComponent(brandSlug)}/locations/${encodeURIComponent(locationSlug)}`,
    { method: "GET" },
    { accessToken, revalidateSeconds: accessToken ? undefined : 60 }
  );
}

/**
 * GET /restaurants/{id}/locations — public, summary shape only. Anonymous
 * callers get the ACTIVE locations only; pass the caller's `accessToken` and
 * the owning owner / an admin additionally get the brand's non-active ones
 * (coming_soon, hidden, closed) — which the owner console needs, or a listing
 * still in setup vanishes from the owner's own card. An authenticated read is
 * never Next-cached (`apiFetch`), so no `revalidateSeconds` alongside a token.
 */
export async function getRestaurantLocations(
  id: number,
  params: PaginationParams = {},
  accessToken?: string
): Promise<PaginatedResponse<LocationSummary>> {
  const query = toQueryString({
    page: params.page,
    page_size: params.page_size,
  });

  return apiFetch<PaginatedResponse<LocationSummary>>(
    `/restaurants/${id}/locations${query}`,
    { method: "GET" },
    { accessToken, revalidateSeconds: accessToken ? undefined : 60 }
  );
}

/** POST /restaurants — auth: owner. */
export async function createRestaurant(
  input: CreateRestaurantInput,
  accessToken: string
): Promise<RestaurantBrand> {
  return apiFetch<RestaurantBrand>(
    "/restaurants",
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** PATCH /restaurants/{id} — auth: owner (must own the brand) or admin. */
export async function updateRestaurant(
  id: number,
  input: UpdateRestaurantInput,
  accessToken: string
): Promise<RestaurantBrand> {
  return apiFetch<RestaurantBrand>(
    `/restaurants/${id}`,
    { method: "PATCH", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * DELETE /restaurants/{id} — auth: admin only. SOFT delete: stamps
 * `deleted_at` and deactivates every active location of the brand in one
 * transaction; the row is kept and the call is idempotent. Returns 204 and
 * never 409s (docs/API_CONTRACTS.md "DELETE /restaurants/{id}").
 */
export async function deleteRestaurant(
  id: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/restaurants/${id}`,
    { method: "DELETE" },
    { accessToken }
  );
}

/**
 * POST /restaurants/{id}/restore — auth: admin only. Clears `deleted_at`.
 * The brand's locations are NOT reactivated (they stay deactivated until
 * re-enabled through their normal status path). Idempotent.
 */
export async function restoreRestaurant(
  id: number,
  accessToken: string
): Promise<RestaurantBrand> {
  return apiFetch<RestaurantBrand>(
    `/restaurants/${id}/restore`,
    { method: "POST" },
    { accessToken }
  );
}

/**
 * POST /restaurants/{id}/follow — auth: registered_user only
 * (docs/API_CONTRACTS.md "Follows (`user_follow`)"). Idempotent server-side
 * — following a brand the caller already follows just returns the existing
 * follow, no error.
 */
export async function followRestaurant(
  id: number,
  accessToken: string
): Promise<FollowOut> {
  return apiFetch<FollowOut>(
    `/restaurants/${id}/follow`,
    { method: "POST" },
    { accessToken }
  );
}

/**
 * DELETE /restaurants/{id}/follow — auth: registered_user only. Idempotent
 * — unfollowing a brand the caller doesn't currently follow is a no-op 204,
 * not an error.
 */
export async function unfollowRestaurant(
  id: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/restaurants/${id}/follow`,
    { method: "DELETE" },
    { accessToken }
  );
}
